src/expr/src/linear.rs

```diff
@@ -1042,3 +1003,42 @@ impl MapFilterProject {
             .project(projection.into_iter().map(|c| remap[&c]));
     }
 }
+
+// A helper method which memoizes expressions by recursively memoizing their parts.
```

Comment:
I'm happy to help here, but I think we want more documentation about what this does. When it was a private method that no one else used, fewer worries, but we'll probably want to spell out e.g. that this places new expressions at the end of `new_scalars`, and .. something about how you should only append to that not `clear` it or anything, and that references need to add `input_arity` etc.